Add lossless tokens

fpringvaldsen opened the pull request:
This PR allows Tokens to keep track of the original text from which they were parsed. It also adds a new `IgnoredWhitespace` token to represent whitespace ignored by the parser that was previously thrown out when tokenizing input. Finally, a `render()` method has been added to the `TokenIterator` class, which reproduces the original input text.

This is the first round of changes to address the new features discussed in https://github.com/typesafehub/config/issues/149.

typesafehub-validator commented:
Hi @fpringvaldsen,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.typesafe.com/contribute/cla](http://www.typesafe.com/contribute/cla)

havocp commented:
This is looking pretty nice!

havocp commented on config/src/main/java/com/typesafe/config/impl/Parser.java:
```diff
@@ -319,6 +319,16 @@ private TokenWithComments nextTokenIgnoringNewline() {
 
+        // Grabs the next Token off of the TokenIterator, ignoring
+        // IgnoredWhitespace tokens
+        private Token nextTokenFromIterator() {
```
if this is called nextTokenIgnoringWhitespace or something it might be a little clearer (fromiterator isn't super relevant to the caller)

havocp commented on config/src/main/java/com/typesafe/config/impl/Tokenizer.java:
```diff
@@ -52,6 +52,14 @@ else if (Character.isISOControl(codepoint))
     }
 
+    static String render(Iterator<Token> tokens) {
+        String renderedText = "";
```
StringBuilder would seem more natural for this maybe?

Passing in an iterable rather than an iterator might also avoid some bugs (since iterator is mutable and caller may forget to make a copy)

fpringvaldsen replied:
@havocp Should we modify `tokenize`, then, to return an iterable rather than an iterator? Or do we expect the user to use the iterator to construct some sort of collection of tokens and pass that into `render`?

havocp replied:
hmm, no that's probably not worth mixing in to this PR. can leave it as is for now.

fpringvaldsen replied:
Alright, I'll leave it as an `Iterator` for now.

havocp commented on config/src/main/java/com/typesafe/config/impl/Tokenizer.java:
```diff
@@ -107,7 +120,12 @@ private Token nextIsASimpleValue(ConfigOrigin baseOrigin,
+                                whitespace.toString());
+                        whitespace.setLength(0);
+                        return t;
+                    }
```
this block of code is in two places, should probably be split into a method

havocp commented on config/src/main/java/com/typesafe/config/impl/Tokenizer.java:
```diff
@@ -478,13 +515,14 @@ private Token pullQuotedString() throws ProblemException {
 
-            return Tokens.newString(lineOrigin, sb.toString());
+            }
+            return Tokens.newString(lineOrigin, sb.toString(), sbOrig.toString());
```
it might be worthwhile to jump through some hoops to try to use the same string twice here, in the common case with no escape sequences, so we don't end up with double the memory. I think that would mean: 1) record triple vs. double quotes flag separately, instead of putting the quotes in the sbOrig string, and 2) check here whether sb.toString.equals(sbOrig.toString).  ?

Now that I say that, I'm fine with leaving it out of this PR though. we can optimize it later.

havocp commented on config/src/main/java/com/typesafe/config/impl/Tokens.java:
```diff
@@ -180,8 +198,8 @@ public int hashCode() {
-        Comment(ConfigOrigin origin, String text) {
-            super(TokenType.COMMENT, origin);
+        Comment(ConfigOrigin origin, String text, boolean doubleSlash) {
+            super(TokenType.COMMENT, origin, (doubleSlash? "//" : "#") + text);
```
this would be a little nicer I think if we do it via class hierarchy; `abstract class Comment extends Token`, `DoubleSlashComment extends Comment`, `HashComment extends Comment`, or something. It saves a tiny bit of ram, but the more important reason is that booleans don't read very well. Later on when you use this boolean in the test suite it's hard to know what it means.  long version at http://blog.ometer.com/2011/01/20/boolean-parameters-are-wrong/

havocp commented on config/src/main/java/com/typesafe/config/impl/Tokens.java:
```diff
@@ -222,7 +240,8 @@ public int hashCode() {
         Substitution(ConfigOrigin origin, boolean optional, List<Token> expression) {
-            super(TokenType.SUBSTITUTION, origin);
+            super(TokenType.SUBSTITUTION, origin,
+                    "${" + (optional? "?" : "") + Tokenizer.render(expression.iterator()) + "}");
```
this makes me wonder if tokenText() should be virtual instead of final, so we don't need to precompute and keep around a field here that's completely derivable from the other fields we already have.

havocp commented on config/src/test/scala/com/typesafe/config/impl/TestUtils.scala:
```diff
@@ -656,6 +657,10 @@ abstract trait TestUtils {
         tokenize(s).asScala.toList
     }
 
+    def tokenizeAsString(s: String) = {
```
this seems like it might belong in TestUtils.scala rather than in the production code

fpringvaldsen replied:
I may be misunderstanding something, but I believe this is in TestUtils.scala.

havocp replied:
ha! easy to fix then :-)

havocp commented on config/src/test/scala/com/typesafe/config/impl/TokenizerTest.scala:
```diff
@@ -33,75 +38,95 @@ class TokenizerTest extends TestUtils {
             tokenKeySubstitution("c.d"), tokenLine(1), Tokens.END)
-        assertEquals(expected, tokenizeAsList(""",:=}{][+="foo"""" + "\"\"\"bar\"\"\"" + """true3.14false42null${a.b}${?x.y}${"c.d"}""" + "\n"))
+        assertEquals(expected, tokenizeAsList(source))
+        assertEquals(source, tokenizeAsString(source))
```
A bunch of these tests could just be using the tokenizerTest method I guess but aren't? Not sure why they weren't already, some historical accident unless I'm missing something.

havocp commented on config/src/test/scala/com/typesafe/config/impl/TokenizerTest.scala:
```diff
@@ -239,15 +286,20 @@ class TokenizerTest extends TestUtils {
+        tokenizerTest(List(tokenInt(10), tokenComment("comment", true)), "10//comment")
+        tokenizerTest(List(tokenInt(10), tokenComment("comment", false)), "10#comment")
+        tokenizerTest(List(tokenDouble(3.14), tokenComment("comment", true)), "3.14//comment")
+        tokenizerTest(List(tokenDouble(3.14), tokenComment("comment", false)), "3.14#comment")
```
here's where `tokenDoubleSlashComment("comment")` and `tokenHashComment("comment")` would be clearer than the boolean

havocp commented on config/src/test/scala/com/typesafe/config/impl/TokenizerTest.scala:
```diff
@@ -239,15 +286,20 @@ class TokenizerTest extends TestUtils {
+        tokenizerTest(List(tokenComment("comment", true), tokenLine(1), tokenComment("comment2", true)),
+                      "//comment\n//comment2")
+        tokenizerTest(List(tokenComment("comment", false), tokenLine(1), tokenComment("comment2", false)),
+                      "#comment\n#comment2")
```
could torture this a little more, like several lines of comments with leading and trailing whitespace, maybe a `\r\n`, etc.

havocp commented:
Looking great, I think/hope all of my comments are minor changes. If any turn out to be hard then we can probably leave them aside for the time being. We should be able to merge this quickly. It may conflict in a trivial way with some of my other PRs today but should be simple to sort out.

fpringvaldsen commented:
@havocp Thanks for the feedback! Looking over your comments it seems like it's all going to be minor changes, I'll begin addressing them this afternoon and hopefully I'll have something ready by the end of the day.

fpringvaldsen commented:
@havocp I've addressed your feedback! Let me know if you have any more.

The main thing I'm wondering about is the new multi-line comment tests I added. Is that what you're looking for? Is the behavior of the `\r` escape character exposed by those tests expected?

havocp commented:
I think the `\r` behavior is correct; it should be treated just like other whitespace I think, which it appears to be here. The HOCON.md spec says that "newline" means only the `\n`.
